core/src/test/java/io/atomix/core/AtomixTest.java

```diff
@@ -246,6 +250,152 @@ public void testRoleChangedListener() throws Exception {
     roleChanged.get();
   }
 
+  @Test
+  public void testAtomixBootstrapPartitions() throws Exception {
+    // given
+    // Partitions \ Nodes
+    //      \   0  1  2
+    //    0     L  F  F
+    //    1     F  L  F
+    //    2     F  F  L
+
+    final List<Map<Integer, RaftServer.Role>> nodeRoles = new ArrayList<>();
+    nodeRoles.add(new HashMap<>());
+    nodeRoles.add(new HashMap<>());
+    nodeRoles.add(new HashMap<>());
+
+    final List<Integer> members = Arrays.asList(1, 2, 3);
+
+    final int firstNodeId = 1;
+    final CompletableFuture<Atomix> nodeOneFuture = setupAtomixWithListener(firstNodeId, members, nodeRoles);
+
+    final int secondNodeId = 2;
+    final CompletableFuture<Atomix> nodeTwoFuture = setupAtomixWithListener(secondNodeId, members, nodeRoles);
+
+    final int thirdNodeId = 3;
+    final CompletableFuture<Atomix> nodeThreeFuture = setupAtomixWithListener(thirdNodeId, members, nodeRoles);
+
+    // then
+    CompletableFuture.allOf(nodeOneFuture, nodeTwoFuture, nodeThreeFuture).join();
+
+    while (nodeRoles.get(0).size() < 3
+        || nodeRoles.get(1).size() < 3
+        || nodeRoles.get(2).size() < 3) {
+      Thread.sleep(100L);
+    }
+
+    final Map<Integer, RaftServer.Role> expectedNodeOneRoles = new HashMap<>();
+    expectedNodeOneRoles.put(1, Role.LEADER);
+    expectedNodeOneRoles.put(2, Role.FOLLOWER);
+    expectedNodeOneRoles.put(3, Role.FOLLOWER);
+    assertEquals(expectedNodeOneRoles, nodeRoles.get(0));
```

Comment:
I think the test can be flaky, running this locally sometimes fail with everyone being a follower (or is that a real bug?).

```java
java.lang.AssertionError: 
Expected :{1=LEADER, 2=FOLLOWER, 3=FOLLOWER}
Actual   :{1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER}
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at io.atomix.core.AtomixTest.testAtomixBootstrapPartitions(AtomixTest.java:291)
```